scripts/TPI.py

```diff
@@ -22,22 +22,30 @@ def solve_tp(g_n_path, omega_S_preTP, rho_s, imm_rates_path, params):
         b_11 = 1.1 * b_ss
         BQ_params = (g_n_path[0], omega_S_preTP, rho_s)
         K_params = (g_n_path[0], omega_S_preTP, imm_rates_path[0, :])
-        BQ_11 = agg.get_BQ(b_11, r_11, BQ_params)
+        BQ_11 = agg.get_BQ(b_11, r_11, BQ_params, method = "TPI")
+        BQ_ss = agg.get_BQ(b_ss, r_ss, BQ_params, method = "SS")
         K_11 = agg.get_K(b_11, K_params)
         BQpath_init = np.zeros(T + S - 1)
         BQpath_init[:T] = np.linspace(BQ_11, BQ_ss, T)
         BQpath_init[T:] = BQ_ss
-        r_11 = firm.get_r(K_11, L_ss, r_params)
-        r_path = firm.get_r(L_ss, r_params)
-        w_path = firm.get_w(L_ss, w_params)
+        L_ss = agg.get_L(n_s, omega_SS)
+        r_11 = firm.get_r(L_ss, K_11, alpha, A, delta)
+        # I can't figure out how r_11 and r_path differ
+        ''' Is r_11 an initial guess? Depending on how you're defining
+        r_11, the arguments passed to the function call above and below
+        will vary
+        ''' 
+        r_path = firm.get_r(L_ss, K_11, alpha, A, delta)
+        w_path = firm.get_w(r_path, alpha, A, delta)
         bmat = np.zeros((S - 1, T + S - 1))
+        # What is b_1 supposed to be?
         bmat[:, 0] = b_1
 
         # Solve for households
```

Comment:
# Solve for partial lifetime decision-makers: (please someone check the periods in here; It's 2 AM in the morning here and I cannot really concentrate very well)
nmat = np.zeros((S, T + S - 1))

for p in range(2, S):
            b_init = np.diag(b_path[S - p + 1:, :p - 1])
            n_init = np.diag(npath[S - p:, :p - 1])
            b_n_init = np.hstack((b_init, n_init))
            b_n_args = (r_path[:p], w_path[:p], beta, sigma, BQpath_init[:p], rho_s[-p:], b_ellipse, upsilon)
            results_b_n = opt.root(hh.FOCs, b_n_init, args=(b_n_args))
            b_solve_p = results_b_n.x[:p - 1]
            n_solve_p = results_b_n.x[p - 1:]
            DiagMaskb = np.eye(p - 1, dtype=bool)
            DiagMaskn = np.eye(p - 1, dtype=bool)

            bmat[S - p:, 1:p] = (DiagMaskb * b_solve_p + bmat[S - p:, 1:p])
            nmat[S - p - 1:, :p - 1] = DiagMaskn * nvec + npath[S - p - 1:, :p-1]

Reply:
I think `.stack` and `.append` are both built-on `.concatenate`. So no real difference to my knowledge. `.append` might flatten an array if axis isn’t specified though. I like your use of `.hstack`.